lighthouse-core/gather/gatherers/forms.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Gatherer = require('./gatherer.js');
+const pageFunctions = require('../../lib/page-functions.js');
+
+/* eslint-env browser, node */
+
+/**
+ * @return {LH.Artifacts['Forms']}
+ */
+/* istanbul ignore next */
+function collectFormElements() {
+  // @ts-ignore - put into scope via stringification
+  const formChildren = getElementsInDocument('textarea, input, label, select'); // eslint-disable-line no-undef
+  const forms = new Map();
+  /** @type { { inputs: LH.Artifacts['FormInputs'][], labels: LH.Artifacts['FormLabels'][] }  } */
+  const formlessObj = {
+    inputs: [],
+    labels: [],
+  };
+  for (const child of formChildren) {
+    const hasForm = child.closest('form');
+    if (hasForm && !forms.has(hasForm)) {
+      const newFormObj = {
+        attributes: {
+          id: hasForm.id,
+          name: hasForm.name,
+          autocomplete: hasForm.autocomplete,
+          // @ts-ignore - put into scope via stringification
+          devtoolsNodePath: getNodePath(hasForm), // eslint-disable-line no-undef
+          // @ts-ignore - put into scope via stringification
+          snippet: getOuterHTMLSnippet(hasForm), // eslint-disable-line no-undef
+        },
+        inputs: [],
+        labels: [],
+      };
+      forms.set(hasForm, newFormObj);
+    }
+    const formObj = hasForm ? forms.get(hasForm) : formlessObj;
+
+    if (child instanceof HTMLInputElement || child instanceof HTMLTextAreaElement) {
+      if (child.type !== 'submit' && child.type !== 'button') {
+        formObj.inputs.push({
+          id: child.id,
+          nodeName: child.nodeName,
+          name: child.name,
+          placeholder: child.placeholder,
+          autocomplete: child.autocomplete,
+          // @ts-ignore - put into scope via stringification
+          devtoolsNodePath: getNodePath(child), // eslint-disable-line no-undef
+          // @ts-ignore - put into scope via stringification
+          snippet: getOuterHTMLSnippet(child), // eslint-disable-line no-undef
+        });
+      }
+    }
+    if (child instanceof HTMLSelectElement) {
```

Comment:
was splitting this out *just* for use of `.type` above?

might be easier since they're otherwise identical to just do

```js
const isButton = child instanceof HTMLInputElement && (child.type === 'submit' || child.type === 'button');
if (isButton) continue;
```

above and merge the two

sidenote: why are we excluding buttons? :)

Reply:
thanks for the suggestion made the edit.
@patrickhulce because its for the autocomplete audits the buttons are not necessary to collect because it wont have an autocomplete attribute

Reply:
ah gotcha so just because the audit won't use them. seems fine until we have a new audit that needs them 👍